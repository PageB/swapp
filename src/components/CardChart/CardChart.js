import React from 'react';
import { useQuery } from '@apollo/react-hooks';
import Chart from 'react-apexcharts';
import PropTypes from 'prop-types';

import { Loading, LoadingError } from '../../components';
import { ALL_STARSHIPS } from '../../queries/starships';
import styles from './CardChart.module.scss';

const CardChart = props => {
  const { data, error, loading } = useQuery(ALL_STARSHIPS);

  if (loading) return <Loading />;
  if (error) return <LoadingError />;

  const {
    allStarships: { edges },
  } = data;

  const filteredStarshipsByClass = edges.filter(
    edge => edge.node.starshipClass === props.starship.starshipClass,
  );

  const findMaxClassValuePerProperty = property => {
    return Math.max(
      ...filteredStarshipsByClass.map(starship =>
        starship.node[property] !== null ? starship.node[property] : 0,
      ),
    );
  };

  const options = {
    labels: ['ML', 'Cost', 'Crew', 'Speed', 'Rating'],
    plotOptions: {
      radar: {
        polygons: {
          strokeColors: '#EFF2F7 ',
          connectorColors: '#EFF2F7',
          fill: {
            colors: ['#000000'],
          },
        },
      },
    },
    dataLabels: {
      style: {
        colors: ['#4BD5EE'],
      },
    },
    chart: {
      background: '#000',
      toolbar: {
        show: false,
      },
    },
    colors: ['#FFE300'],
    markers: {
      size: 4,
      colors: ['#fff'],
      strokeColor: '#FF4560',
      strokeWidth: 2,
    },
    yaxis: {
      show: false,
      tickAmount: 5,
    },
  };
  const series = [
    {
      name: props.starship.name,
      data: [
        props.starship.maxMLPerHour !== null ? props.starship.maxMLPerHour : 0,
        props.starship.cost !== null ? props.starship.cost : 0,
        props.starship.crew !== null ? props.starship.crew : 0,
        props.starship.maxAtmosphericSpeed !== null ? props.starship.maxAtmosphericSpeed : 0,
        props.starship.hyperdriveRating !== null ? props.starship.hyperdriveRating : 0,
      ],
    },
    {
      name: 'Starship Class Max',
      data: [
        findMaxClassValuePerProperty('maxMLPerHour'),
        findMaxClassValuePerProperty('cost'),
        findMaxClassValuePerProperty('crew'),
        findMaxClassValuePerProperty('hyperdriveRating'),
        findMaxClassValuePerProperty('maxAtmosphericSpeed'),
      ],
    },
  ];

  return (
    <div className={[styles.Chart, styles[props.theme]].join(' ')}>
      <Chart options={options} series={series} type="radar" height="350" />
    </div>
  );
};

CardChart.propTypes = {
  theme: PropTypes.oneOf(['DarkTheme', 'LightTheme']),
};

CardChart.defaultProps = {
  theme: 'DarkTheme',
};

export default CardChart;
