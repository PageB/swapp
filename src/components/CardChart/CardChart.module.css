.Chart {
  width: 280px;
}
