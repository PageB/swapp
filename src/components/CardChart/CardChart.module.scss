@import '../../colors.scss';
@import '../../constants.scss';

.Chart {
  width: $card-width;
}

.DarkTheme {
  color: $vivid-yellow;
}

.LightTheme {
  color: $crayola;
}
