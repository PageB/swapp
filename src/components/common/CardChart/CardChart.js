import React, { Component } from 'react';
import Chart from 'react-apexcharts';

import styles from './CardChart.module.scss';

class CardChart extends Component {
  constructor(props) {
    super(props);
    
    this.state = {
      options: {
        plotOptions: {
          radar: {
            polygons: {
              strokeColors: '#EFF2F7 ',
              connectorColors: '#EFF2F7',
              fill: {
                colors: ['#000000'],
              },
            },
          },
        },
        yaxis: {
          show: false,
          max: 500,
        },
        dataLabels: {
          style: {
            colors: ['#4BD5EE'],
          },
        },
        chart: {
          background: '#000',
          toolbar: {
            show: false,
          },
        },
        colors: ['#FFE300'],
        labels: ['Class', 'Cost', 'Crew', 'Speed', 'Rating'],
      },
      series: [
        {
          name: props.name,
          data: [80, 50, 30, 40, 100],
        },
      ],
    };
  }

  render() {
    return (
      <div className={[styles.Chart, styles[this.props.theme]].join(' ')}>
        <p>{this.state.name}</p>
        <Chart options={this.state.options} series={this.state.series} type="radar" height="350" />
      </div>
    );
  }
}

export default CardChart;
