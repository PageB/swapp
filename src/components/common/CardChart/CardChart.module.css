.Chart {
  width: 280px;
}

.DarkTheme {
  color: #ffe300;
}

.LightTheme {
  color: #4bd5ee;
}

.DarkTheme .apexcharts-canvas {
  background: yellow;
}