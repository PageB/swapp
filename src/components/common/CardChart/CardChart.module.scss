@import '../../../colors.scss';

.Chart {
  width: 280px;
}

.DarkTheme {
  color: $vivid-yellow;
}

.LightTheme {
  color: $crayola;
}

.DarkTheme .apexcharts-canvas {
  background: yellow;
}
