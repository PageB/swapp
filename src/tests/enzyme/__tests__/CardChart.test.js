import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import { starWarsCharacters } from '../../../mock/data';

import CardChart from '../../../components/CardChart/CardChart';

configure({ adapter: new Adapter() });

describe('CardChart Component', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<CardChart starship={starWarsCharacters[0].starships[0]} />);

    expect(wrapper).toMatchSnapshot();
  });
});
