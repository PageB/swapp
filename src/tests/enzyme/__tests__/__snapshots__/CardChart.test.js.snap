// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardChart Component should match snapshot 1`] = `
<div
  className="Chart DarkTheme"
>
  <p />
  <r
    height="350"
    options={
      Object {
        "chart": Object {
          "background": "#000",
          "toolbar": Object {
            "show": false,
          },
        },
        "colors": Array [
          "#FFE300",
        ],
        "dataLabels": Object {
          "style": Object {
            "colors": Array [
              "#4BD5EE",
            ],
          },
        },
        "labels": Array [
          "ML",
          "Cost",
          "Crew",
          "Speed",
          "Rating",
        ],
        "plotOptions": Object {
          "radar": Object {
            "polygons": Object {
              "connectorColors": "#EFF2F7",
              "fill": Object {
                "colors": Array [
                  "#000000",
                ],
              },
              "strokeColors": "#EFF2F7 ",
            },
          },
        },
        "yaxis": Object {
          "max": 3000,
          "show": false,
        },
      }
    }
    series={
      Array [
        Object {
          "data": Array [
            631,
            631,
            108,
            1159,
            150,
          ],
          "name": undefined,
        },
      ]
    }
    type="radar"
    width="100%"
  />
</div>
`;
